In [1]:
import pandas as pd

# Load the Titanic dataset if it's not already
df = pd.read_csv("Titanic Dataset.csv")

print("✓ Data loaded successfully!")
print(f"Original shape: {df.shape}")
print(f"\nOriginal columns: {df.columns.tolist()}")

✓ Data loaded successfully!
Original shape: (1309, 14)

Original columns: ['pclass', 'survived', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket', 'fare', 'cabin', 'embarked', 'boat', 'body', 'home.dest']


# Task 1

In [6]:
print(f"Rows: {len(df)}")

print(f"Columns: {len(df.columns)}")

Rows: 1309
Columns: 14


In [7]:
# Displaying first five rows

print(df.head())

   pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  


In [8]:
# Checking for missing values in original data

print("Missing values in the original data:")
print(df.isnull().sum())

Missing values in the original data:
pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64


Which columns have the most missing values?

Body.

# Task 2

In [10]:
df = df[['pclass','survived','sex','age','sibsp','parch','fare','embarked']]

print("✓ Columns selected!")
print(f"New shape: {df.shape}")
print(f"\nColumns: {df.columns.tolist()}")

✓ Columns selected!
New shape: (1309, 8)

Columns: ['pclass', 'survived', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked']


Why did we drop the 'name' column?

isnt important

Why did we drop the 'cabin' column?

doesnt help

Why did we drop the 'boat' column?

doesnt help

# Task 3

In [11]:
# Checking for missing values in the NEW dataset

print("Missing values after feature selection:")
print(df.isnull().sum())

Missing values after feature selection:
pclass        0
survived      0
sex           0
age         263
sibsp         0
parch         0
fare          1
embarked      2
dtype: int64


Q: How many missing values are in the 'age' column?

A: 263

Q: How many missing values are in the 'fare' column?

A: 1

Q: How many missing values are in the 'embarked' column?

A: 2

# Task 4

In [13]:
# Calculating median age
median_age = df['age'].median()
print(f"Median age: {median_age}")

Median age: 28.0


In [15]:
# Filling in the missing age values with the median

# df['age'].fillna(median_age, inplace=True) ---> This one apparently gives you a future warning, so the code below is the more "improved version", though
# it does the same task.
df.fillna({'age': median_age}, inplace=True)

print("✓ Missing ages filled with median!")
print(f"Missing ages now: {df['age'].isnull().sum()}")

✓ Missing ages filled with median!
Missing ages now: 0


In [16]:
# Calculate both median and mean
print(f"Median age: {df['age'].median():.2f}")
print(f"Mean age: {df['age'].mean():.2f}")

Median age: 28.00
Mean age: 29.50


Q: Which is larger, the median or mean? Why might this be?

A: The mean turns out to be the larger value between the two. This most likely occured because most of the missing values are now filled up with that median age value. The mean though would be quite different in comparison to the previous mean we had with those missing values. The mean value is now drastically changed since those missing values have been filled.

# Task 5

In [17]:
median_fare = df['fare'].median()
print(f"Median fare: ${median_fare:.2f}")

df.fillna({'fare': median_fare}, inplace=True)

print("✓ Missing fare filled!")
print(f"Missing fares now: {df['fare'].isnull().sum()}")

Median fare: $14.45
✓ Missing fare filled!
Missing fares now: 0


# Task 6

In [19]:
rows_before = len(df)
print(f"Rows before dropping: {rows_before}")

# Dropping rows with missing embarked values

df.dropna(subset=['embarked'], inplace=True)

rows_after = len(df)
rows_dropped = rows_before - rows_after

print("✓ Rows with missing embarked dropped!")
print(f"Rows after dropping: {rows_after}")
print(f"Rows dropped: {rows_dropped}")
print(f"Missing embarked now: {df['embarked'].isnull().sum()}")

Rows before dropping: 1307
✓ Rows with missing embarked dropped!
Rows after dropping: 1307
Rows dropped: 0
Missing embarked now: 0


Q: What percentage of data did we lose by dropping these rows?

A: According to the code, there was technically no data lost, since no rows were dropped.

# Task 7

In [20]:
# Check for any remaining missing values

print("Final missing value check:")
print(df.isnull().sum())
print(f"\nTotal missing values: {df.isnull().sum().sum()}")

Final missing value check:
pclass      0
survived    0
sex         0
age         0
sibsp       0
parch       0
fare        0
embarked    0
dtype: int64

Total missing values: 0


No missing values!

# Task 8

In [21]:
# Display summary statistics

print("Summary statistics after cleaning:")
df.describe()

Summary statistics after cleaning:


pclass     survived          age        sibsp        parch  \
count  1307.000000  1307.000000  1307.000000  1307.000000  1307.000000   
mean      2.296863     0.381025    29.471821     0.499617     0.385616   
std       0.836942     0.485825    12.881592     1.042273     0.866092   
min       1.000000     0.000000     0.170000     0.000000     0.000000   
25%       2.000000     0.000000    22.000000     0.000000     0.000000   
50%       3.000000     0.000000    28.000000     0.000000     0.000000   
75%       3.000000     1.000000    35.000000     1.000000     0.000000   
max       3.000000     1.000000    80.000000     8.000000     9.000000   

              fare  
count  1307.000000  
mean     33.209595  
std      51.748768  
min       0.000000  
25%       7.895800  
50%      14.454200  
75%      31.275000  
max     512.329200

Q: What is the average age after filling missing values?

A: 29 (29.4)

Q: What is the average fare?

A: 33 (33.2)

# Task 9

In [22]:
# Save cleaned data
# Writes the DataFrame to a new CSV file without adding extra row numbers (index)
df.to_csv("Titanic_Cleaned.csv", index=False)     

# Prints a confirmation message so the user knows the save was successful
print("✓ Cleaned data saved to 'Titanic_Cleaned.csv'")

✓ Cleaned data saved to 'Titanic_Cleaned.csv'


Reflection Questions
Answer these questions based on your work:

1. Why is it important to check for missing values before building a model?

Answer: helps add for a more accurate model later

2. When should you fill missing values vs. drop rows?

Answer: It depends on how many missing values there are then they should be filled.

3. Why did we use median instead of mean to fill missing ages?

Answer: to ensure a more stable estimate for filling in missing ages, the median should be used instead of the mean, which can be easily changed by extreme big values.

4. What could happen if we trained a model on data with missing values?

Answer: bad info with the model would be produced

5. Name one real-world scenario where missing data might occur.

Answer: plane Crashes

